thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/DetectionConfigTuner.java

```diff
@@ -0,0 +1,158 @@
+    long configId = detectionConfig == null ? 0 : detectionConfig.getId();
+    String componentClassName = componentProps.get(PROP_CLASS_NAME).toString();
+    Map<String, Object> yamlParams = ConfigUtils.getMap(componentProps.get(PROP_YAML_PARAMS));
+    InputDataFetcher dataFetcher = new DefaultInputDataFetcher(dataProvider, configId);
+    Tunable tunable = instantiateTunable(componentClassName, yamlParams, dataFetcher);
+
+    // round to daily boundary
+    DateTimeZone timezone = DateTimeZone.forID(this.datasetConfig.getTimezone() == null ? DEFAULT_TIMEZONE : this.datasetConfig.getTimezone());
+    DateTime start = new DateTime(startTime, timezone).withTimeAtStartOfDay();
+    DateTime end =  new DateTime(endTime, timezone).withTimeAtStartOfDay();
+    Interval window = new Interval(start, end);
+
+    // TODO can you use componentProps instead? what is the difference btw component and componentSpec
+    Map<String, Object> existingComponentSpec =
+        detectionConfig.getComponents().containsKey(componentKey) ?
+            MapUtils.getMap(detectionConfig.getComponents(), componentKey) : Collections.emptyMap();
+
+    // TODO: if dimension drill down applied, pass in the metric urn of top dimension
+    return tunable.tune(existingComponentSpec, window, this.metricUrn);
+  }
+
+  private Tunable instantiateTunable(String componentClassName, Map<String, Object> yamlParams, InputDataFetcher dataFetcher)
+      throws Exception {
+    String tunableClassName = DETECTION_REGISTRY.lookupTunable(componentClassName);
+    Class clazz = Class.forName(tunableClassName);
+    Class<AbstractSpec> specClazz = (Class<AbstractSpec>) Class.forName(getSpecClassName(clazz));
+    AbstractSpec spec = AbstractSpec.fromProperties(yamlParams, specClazz);
+    Tunable tunable = (Tunable) clazz.newInstance();
+    tunable.init(spec, dataFetcher);
+    return tunable;
+  }
+
+  /**
+   * Scans through all the tunable components and injects tuned model params into the detection config
+   */
+  public DetectionConfigDTO tune(long tuningWindowStart, long tuningWindowEnd) {
+    Map<String, Object> allComponentSpecs = this.detectionConfig.getComponentSpecs();
+    for (Map.Entry<String, Object> componentSpec : allComponentSpecs.entrySet()) {
+      String componentKey = componentSpec.getKey();
+      Map<String, Object> componentProps = ConfigUtils.getMap(componentSpec.getValue());
+
+      // For tunable components, the model params are computed from user supplied yaml params and previous model params.
+      String componentClassName = componentProps.get(PROP_CLASS_NAME).toString();
+      String type = DetectionUtils.getComponentType(componentKey);
+      if (!TURNOFF_TUNING_COMPONENTS.contains(type) && DETECTION_REGISTRY.isTunable(componentClassName)) {
+        try {
+          componentProps.putAll(getTunedSpecs(componentKey, componentProps, tuningWindowStart, tuningWindowEnd));
+          detectionConfig.setComponentSpecs(componentProps);
```

Comment:
is `componentProps` the properties for one component? should we set it as  `setComponentSpecs` of the whole detection config?